### PR TITLE
Deduper: Fix Hash-dependent Tests

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
- #3254 upgraded the `prio-graph` library. 
- That library uses `ahash v0.8.11`.
- Cargo will use highest non-breaking version, in this case `0.8.11`
- Current `Deduper` tests depend on consistent hashing output, which is explicitly not guaranteed by the `ahash` crate

#### Summary of Changes
- Refactor `Deduper` to have a fn to operate on already hashed values
- Remove tests on `dedup` itself, instead testing `dedup_hashed`

Fixes #

### CODE REVIEW

**Comment by apfitzge** [comment]
Refactor does not have significant impact on performance:

```
master:
test bench_dedup_baseline           ... bench:          19.71 ns/iter (+/- 0.42)
test bench_dedup_diff_big_packets   ... bench:     451,904.20 ns/iter (+/- 12,519.65)
test bench_dedup_diff_small_packets ... bench:      84,986.67 ns/iter (+/- 6,705.59)
test bench_dedup_reset              ... bench:      33,601.90 ns/iter (+/- 6,465.13)
test bench_dedup_same_big_packets   ... bench:     399,453.12 ns/iter (+/- 3,142.10)
test bench_dedup_same_small_packets ... bench:      71,592.62 ns/iter (+/- 616.06)

branch:
test bench_dedup_baseline           ... bench:          19.69 ns/iter (+/- 0.28)
test bench_dedup_diff_big_packets   ... bench:     455,470.80 ns/iter (+/- 12,246.42)
test bench_dedup_diff_small_packets ... bench:      89,647.91 ns/iter (+/- 4,498.73)
test bench_dedup_reset              ... bench:      33,571.86 ns/iter (+/- 6,420.38)
test bench_dedup_same_big_packets   ... bench:     413,564.57 ns/iter (+/- 3,834.36)
test bench_dedup_same_small_packets ... bench:      74,867.61 ns/iter (+/- 1,109.70)
```

File: perf/src/deduper.rs
Comment (behzadnouri) [inline perf/src/deduper.rs]: This change is reducing test coverage by a lot.
Also I think the existing `test_case`s provides a clear explicit demonstration of how the deduper is expected to work.
There is also in fact some value in these hard-coded values because it helps to catch unexpected changes in behavior of the deduper after an upgrade or on a certain machine, etc.

Does updating the values inside `test_case`s unblock you for https://github.com/anza-xyz/agave/pull/3254?
If so, I think the best is to just update these and keep the test coverage as before. 

If someone still complained that the test failed on their machine, we can think of a different solution which does not reduce coverage.

Comment (apfitzge) [reply perf/src/deduper.rs]: What cases are not covered? We're not running a bunch of random loops so there's less data going through them, but we're instead targetting specific cases for the tests.

> There is also in fact some value in these hard-coded values because it helps to catch unexpected changes in behavior of the deduper after an upgrade or on a certain machine, etc.

Sure there's value in testing the actual deduping, but there's not value in testing the actual hash values.

We could generated random (seeded) u64s for our hashes, and do similar large loops with varied parameters and make sure the values do not change, wdyt?
But leaving the tests as is, i.e. relying on consistent hashes for a library that doesn't guarantee that, is not a good idea imo.

Comment (behzadnouri) [reply perf/src/deduper.rs]: > but there's not value in testing the actual hash values

I disagree with that. If after an update to ahash or on a certain machine etc we get too many false positive, the hard-coded values helps us to notice that.

> relying on consistent hashes for a library that doesn't guarantee that, is not a good idea

We need to guarantee that deduper has sane behavior.
If an update to the library breaks that guarantee, then we need some way to notice that and perhaps avoid the library update, or use a different hasher, etc. 

I still think if updating these values unblocks you for https://github.com/anza-xyz/agave/pull/3254 then the best is to just do that.

Comment (apfitzge) [reply perf/src/deduper.rs]: > I disagree with that. If after an update to ahash or on a certain machine etc we get too many false positive, the hard-coded values helps us to notice that.

If we want this to be guaranteed, we should have separate tests for that. That rely on checking collision rate over many elements, not exact hash value.

Comment (behzadnouri) [reply perf/src/deduper.rs]: Lets say ahash has a malicious update which always returns 42 no matter what the input to the hasher is.
Obviously that breaks the deduper, and the existing test catches that and prevents us from updating to that malicious version of the library.

I am making an extreme example to make the point across but I am sure you can think of more subtle ways a library can  rug pull us, in which case such hard-coded tests will be useful.

Comment (apfitzge) [reply perf/src/deduper.rs]: and I'm saying then let's add a separate test that tests JUST collision rate, which is the concern.

The deduper tests are fragile because of the current structure relying on exact hash value consistency. We know that isn't there so we should not rely on it.

Much more than deduper is at risk if a hash had a malicious update like that.
A distinct test makes it very obvious if such a regression occurs.
The current tests just makes people update hard coded constants without any understanding of what is actually wrong

Comment (apfitzge) [reply perf/src/deduper.rs]: To be clear I will add these tests. I'm not going to merge this PR as is.

But I want to just get buy-in from you before I go out and write them

Comment (behzadnouri) [reply perf/src/deduper.rs]: I think you are just repeating previous statements and if I want to argue I have to repeat my previous responses.

>  The current tests just makes people update hard coded constants without any understanding of what is actually wrong

If the reviewer has enough context on why that test is there, he can catch problematic updates.

Comment (apfitzge) [reply perf/src/deduper.rs]: I pushed the test(s) with removed asserts on exact pop count and dup count since those are the problem with inconsistent hashing. The tests simply check the false postive rate.

This, I hope, resolves your concerns about hashing quality regression, and I think this is reasonable.

Comment (behzadnouri) [reply perf/src/deduper.rs]: > removed asserts on exact pop count and dup count

This still means reducing test coverage and I frankly do not like that.

I am updating ahash to 0.8.11 in https://github.com/anza-xyz/agave/pull/3267.
If all is green there I will merge that and you can rebase.

That will unblock your work without reducing test-coverage so I can sleep well at night.

Comment (apfitzge) [reply perf/src/deduper.rs]: I'm not trying to simply unblock my work. That's just how I found the issue. I'm trying to improve tests because from my perspective they are badly designed.

Comment (behzadnouri) [reply perf/src/deduper.rs]: My point is that I am not convinced this change is an improvement to the test.

We have to guarantee that the deduper still behaves reasonably even if the dependencies don't provide certain guarantees. The test was intentionally written like that to assert that.

File: perf/src/deduper.rs
Comment (alessandrod) [inline perf/src/deduper.rs:85]: nit: this can be Iterator doesn't have to be ExactSizeIterator

**Review by alessandrod — DISMISSED** [review]
This looks good to me.

I think that test_dedup_hashed (check dedups with predefined hashes and expected values), plus test_dedup_seeded (smoke test + check that false positive rate makes sense) is enough to test that dedup works.

I don't think that testing hardcoded ahash values is worth the effort of having to maintain them - the fact that this PR exists is proof of it. 

I don't buy the argument "what if ahash gets taken over". We have nearly 5000 deps in the dep tree. We can't afford to individually test them all.

**Review by behzadnouri — CHANGES_REQUESTED** [review]
ahash is already upgraded to v0.8.11 in https://github.com/anza-xyz/agave/pull/1170. So https://github.com/anza-xyz/agave/pull/3254 should be unblocked; and I am frankly _not_ comfortable to reduce the test coverage here.


> I don't think that testing hardcoded ahash values is worth the effort of having to maintain them

Below is all the effort that it takes. Is that really too much?
https://github.com/anza-xyz/agave/pull/1170/files#diff-cdd6f30ed42d62cdc87d3ad94d01f62383b87ab56ef8c5065ed6eb784bea1f82

**Comment by alessandrod** [comment]
> and I am frankly not comfortable to reduce the test coverage here.

can you point out exactly where coverage is reduced? What is not getting tested?

> Below is all the effort that it takes. Is that really too much?

The effort is understanding why things break, making the PR, the comments and the disagreements.

**Comment by alessandrod** [comment]
master coverage

image

PR coverage

image

we're not testing ahash, we're testing explicit hash values in test_dedup_hashed, which imo is an improvement

**Comment by behzadnouri** [comment]
> > and I am frankly not comfortable to reduce the test coverage here.
> 
> can you point out exactly where coverage is reduced? What is not getting tested?

https://github.com/anza-xyz/agave/pull/3263#discussion_r1811203425

> > Below is all the effort that it takes. Is that really too much?
> 
> The effort is understanding why things break, making the PR, the comments and the disagreements.

As someone who had to several times debug local cluster tests failures after very unrelated changes, the effort over resolving this particular test failure is nothing in comparison.
I would have also loved to use that same argument to delete all local cluster tests but I do realize that might cause some folks to lose sleep at night, so I wouldn't do that.

**Comment by behzadnouri** [comment]
@alessandrod I reverted your edits to my comments.
I am not sure if the edits were intentional or not, but either way, it makes reading through the conversation pretty confusing and besides that I didn't get a notification that there were new comments. Adding a screenshot of the edits below just to preserve the conversation.

That said, ahash is already upgraded to v0.8.11 in https://github.com/anza-xyz/agave/pull/1170, and there is nothing blocking https://github.com/anza-xyz/agave/pull/3254. So I hope we can close this PR and instead move on to more important stuff.

If the current deduper test is too paranoid or not adding value that is a totally fine opinion. But I disagree with that premise, so I would like to keep the test as is and I will take the responsibility to maintain it so it wouldn't burden others in their work.  


editss

**Comment by alessandrod** [comment]
> @alessandrod I reverted your edits to my comments. I am not sure if the edits were intentional or not, but either way, it makes reading through the conversation pretty confusing and besides that I didn't get a notification that there were new comments. Adding a screenshot of the edits below just to preserve the conversation.

Omg sorry about that, I meant to comment, I don't know how it ended up being an edit 🤦‍♂️
 
> If the current deduper test is too paranoid or not adding value that is a totally fine opinion. But I disagree with that premise, so I would like to keep the test as is and I will take the responsibility to maintain it so it wouldn't burden others in their work.

Do you agree that a test that breaks anytime a dep is upgraded - even accidentally, like in the case that motivated this PR - creates more work and keeps us from more important work?

Anyway, I'll leave to @apfitzge to decide because it's his PR. My opinion is that his test changes are an improvement and should be merged.

**Comment by apfitzge** [comment]
>  I will take the responsibility to maintain it so it wouldn't burden others in their work.

This is what I take issue with though. If my unrelated PR can break the deduper tests now I **am** already burdened. I am blocked waiting for the maintainer of said tests to unblock me.

I'm not going to push more. Ultimately these are your tests, so if you still want me to close it I will do so. But I do hope you re-consider the next time it breaks - and it will.

**Comment by behzadnouri** [comment]
> > If the current deduper test is too paranoid or not adding value that is a totally fine opinion. But I disagree with that premise, so I would like to keep the test as is and I will take the responsibility to maintain it so it wouldn't burden others in their work.
> 
> Do you agree that a test that breaks anytime a dep is upgraded - even accidentally, like in the case that motivated this PR - creates more work and keeps us from more important work?

* With that same argument, anytime I have a CI failure I can go remove `assert`s from any failing test which I do not like or I do not understand.
* The test was intentionally written so to explicitly catch changes of behavior in the hasher that are not suitable for the deduper: [#3263 (comment)](https://github.com/anza-xyz/agave/pull/3263#discussion_r1811128102).
* It will not cause any burden if folks don't start off by assuming that the test is wrong and try to fix it. Just forward the CI failure to whoever `git blame` points to and this would have been resolved in 2 minutes.


> > I will take the responsibility to maintain it so it wouldn't burden others in their work.
> 
> This is what I take issue with though. If my unrelated PR can break the deduper tests now I **am** already burdened. I am blocked waiting for the maintainer of said tests to unblock me.

Your change does not count as unrelated PR though.
Your change was pulling in a new version of the ahash dependency, and the test was intentionally written so to explicitly catch changes of behavior in the hasher that are not suitable for the deduper: [#3263 (comment)](https://github.com/anza-xyz/agave/pull/3263#discussion_r1811128102).


I feel like I am just repeating myself and I really do not like to keep doing that.
Lets move on here and keep the test as is, and if you had issues again with this test (or any of my other tests) please just forward to me and I will address whatever the issue is.

**Comment by apfitzge** [comment]
Disagree with just about everything you said, but I don't care enough. Will just bug you **when** this breaks again.

**Pull request closed without merging**
